libraries/Microsoft.Bot.Builder.Dialogs/SkillDialog.cs

```diff
@@ -197,5 +208,66 @@ private async Task<Activity> SendToSkillAsync(ITurnContext context, Activity act
+        {
+            if (activity?.Attachments != null && !string.IsNullOrWhiteSpace(connectionName))
+            {
+                foreach (var attachment in activity.Attachments.Where(a => a?.ContentType == OAuthCard.ContentType))
+                {
+                    var oauthCard = ((JObject)attachment.Content).ToObject<OAuthCard>();
+                    if (!string.IsNullOrWhiteSpace(oauthCard.TokenExchangeResource?.Uri))
+                    {
+                        var tokenExchangeProvider = turnContext.Adapter as IExtendedUserTokenProvider;
+                        if (tokenExchangeProvider == null)
+                        {
+                            return false;
+                        }
+
+                        TokenResponse result;
+                        try
+                        {
+                            result = await tokenExchangeProvider.ExchangeTokenAsync(
+                               turnContext,
+                               connectionName,
+                               turnContext.Activity.From.Id,
+                               new TokenExchangeRequest(oauthCard.TokenExchangeResource.Uri)).ConfigureAwait(false);
+                        }
+                        catch
+                        {
+                            return false;
+                        }
+
+                        if (!string.IsNullOrWhiteSpace(result?.Token))
+                        {
+                            // Send an invoke back to the skill
+                            return await SendTokenExchangeInvokeToSkillAsync(activity, oauthCard.TokenExchangeResource.Id, oauthCard.ConnectionName, result.Token, cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private async Task<bool> SendTokenExchangeInvokeToSkillAsync(Activity incomingActivity, string id, string connectionName, string token, CancellationToken cancellationToken)
+        {
+            var activity = incomingActivity.CreateReply() as Activity;
+            activity.Type = ActivityTypes.Invoke;
+            activity.Name = SignInConstants.TokenExchangeOperationName;
+            activity.Value = new TokenExchangeInvokeRequest()
+            {
+                Id = id,
+                Token = token,
+                ConnectionName = connectionName
+            };
+
+            // route the activity to the skill
+            var skillInfo = DialogOptions.Skill;
+            var response = await DialogOptions.SkillClient.PostActivityAsync<ExpectedReplies>(DialogOptions.BotId, skillInfo.AppId, skillInfo.SkillEndpoint, DialogOptions.SkillHostEndpoint, incomingActivity.Conversation.Id, activity, cancellationToken).ConfigureAwait(false);
+
+            // Check response status: true if success, false if failure
+            return response.Status >= 200 && response.Status <= 299;
```

Comment:
Shouldn't we throw an exception here?

Reply:
For 5xx, do we retry? Returning false doesn't seem to be the right stance to take.

For 4xx we fucked up and there's nothing to do. Throw an exception.

Reply:
throwing an exception breaks the fallback path . Token exchange failures is not a fatal error and the application just needs to show the oauth card

Reply:
There's no retry for token exchange failures